<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>XeniumSpatialAnalysisWorkshop • XeniumSpatialAnalysisWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="XeniumSpatialAnalysisWorkshop">
<meta property="og:description" content="This workflow package shows how a basic analysis of spatial transcriptomics datasets is performed to gain biological insight.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">XeniumSpatialAnalysisWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/workflow_spatial_xenium.html">Introduction to imaging-based spatial transcriptomics analysis (10x Xenium breast cancer)</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bhuvad/XeniumSpatialAnalysisWorkshop">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="introduction-to-imaging-based-spatial-transcriptomics-analysis">Introduction to imaging-based spatial transcriptomics analysis<a class="anchor" aria-label="anchor" href="#introduction-to-imaging-based-spatial-transcriptomics-analysis"></a>
</h1></div>
<p>Biological systems are complex and heterogeneous, involving a diverse suite of cells with unique functions, interacting and communicating to produce complex behaviour at the tissue-level. Understanding such complexity requires detailed measurements that are now made possible through emerging sub-cellular resolution spatial molecular technologies. These technologies allow us to measure the activity of 100s of genes across &gt;100,000 cells from a biological tissue. In the context of cancer, these technologies allow us to study how cancer cells interact with their environment to survive. To unleash the power of these technologies, we need appropriate bioinformatics analysis to generate biological insight.</p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this workshop, we will analyse a sub-sampled spatial transcriptomics dataset to demonstrate quality control, normalisation, cell typing, spatial domain identification, and domain-specific functional analysis. We will use spatially aware computational methods to perform analyses where such methods are available. At the end of the workshop, attendees will be equipped with the computational tools and data structures used to analyse spatial transcriptomics datasets. They will also be able to analyse their own spatial transcriptomics datasets and decipher complex behaviour in their biological systems of interest.</p>
</div>
<div class="section level2">
<h2 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h2>
<p>This workshop will be relevant to anyone interested in analysing data from emerging high-resolution imaging-based spatial molecular technologies such as 10x Xenium, NanoString CosMx, and Vizgen MERSCOPE.</p>
<p>Attendees will require a laptop with internet access and should have familiarity with R, some familiarity with Bioconductor. Some basic knowledge on spatial transcriptomics technologies and the standard steps needed to gain insight from the data they generate is desirable.</p>
</div>
<div class="section level2">
<h2 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h2>
<table class="table">
<colgroup>
<col width="91%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction &amp; setup</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Part 1. Preprocess spatial `omics data (QC and Normalisation)</td>
<td>45m</td>
</tr>
<tr class="odd">
<td>Part 2. Infer spatial biology (cell typing, spatial domains)</td>
<td>45m</td>
</tr>
<tr class="even">
<td>Part 3. Interpret spatial biology (DE, GSEA)</td>
<td>45m</td>
</tr>
<tr class="odd">
<td>Q &amp; A</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h2>
<div class="section level3">
<h3 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h3>
<ul>
<li>Preprocess spatial transcriptomics datasets.</li>
<li>Infer cell types, spatial domains, and biological function from spatial transcriptomics datasets.</li>
<li>Understand the importance of visualisation in bioinformatics and computational biology.</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>Exectue a standard spatial transcritpomics analysis pipeline.</li>
<li>Interpret spatial biology from spatial transriptomics datasets.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="workshop-package-installation">Workshop package installation<a class="anchor" aria-label="anchor" href="#workshop-package-installation"></a>
</h2>
<div class="section level3">
<h3 id="guide">Guide<a class="anchor" aria-label="anchor" href="#guide"></a>
</h3>
<p>This is necessary in order to reproduce the code shown in the workshop. The workshop is designed for R <code>4.4.1</code> and Bioconductor <code>3.20</code>, and can be installed using one of the two ways below.</p>
</div>
<div class="section level3">
<h3 id="via-docker-image">Via Docker image<a class="anchor" aria-label="anchor" href="#via-docker-image"></a>
</h3>
<p>If you’re familiar with <a href="https://docs.docker.com/get-docker/" class="external-link">Docker</a> you could use the Docker image which has all the software pre-configured to the correct versions.</p>
<pre><code>docker run -e PASSWORD=password -p 8787:8787 bhuvad/xeniumspatialanalysisworkshop:latest</code></pre>
<p>Once running, navigate to <a href="http://localhost:8787/" class="external-link uri">http://localhost:8787/</a> and then login with <code>Username:rstudio</code> and <code>Password:password</code>.</p>
<p>You should see the Rmarkdown file with all the workshop code which you can run.</p>
</div>
<div class="section level3">
<h3 id="via-github">Via GitHub<a class="anchor" aria-label="anchor" href="#via-github"></a>
</h3>
<p>Alternatively, you could install the workshop using the commands below in R <code>&gt;4.4.1</code> and BioConductor <code>&gt;3.20</code>.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'remotes'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install workshop package</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"bhuvad/XeniumSpatialAnalysisWorkshop"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># To view vignettes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bhuvad.github.io/XeniumSpatialAnalysisWorkshop/" class="external-link">XeniumSpatialAnalysisWorkshop</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html" class="external-link">browseVignettes</a></span><span class="op">(</span><span class="st">"XeniumSpatialAnalysisWorkshop"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bhuvad/XeniumSpatialAnalysisWorkshop/">Browse source code</a></li>
<li><a href="https://github.com/bhuvad/XeniumSpatialAnalysisWorkshop/issues">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing XeniumSpatialAnalysisWorkshop</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Dharmesh D. Bhuva <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-6398-9157" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Chin Wee Tan <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-9695-7218" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Dharmesh D. Bhuva, Chin Wee Tan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
